cmd/zed/agents/zfs_retire.c

```diff
@@ -318,8 +337,21 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
 		    &vdev)) == NULL)
 			return;
 
-		if (fmd_prop_get_int32(hdl, "spare_on_remove"))
-			replace_with_spare(hdl, zhp, vdev);
+		devname = zpool_vdev_name(NULL, zhp, vdev, B_FALSE);
+
+		/* Can't replace l2arc with a spare: offline the device */
+		if (nvlist_lookup_string(nvl, FM_EREPORT_PAYLOAD_ZFS_VDEV_TYPE,
+		    &devtype) == 0 && strcmp(devtype, VDEV_TYPE_L2CACHE) == 0) {
+			fmd_hdl_debug(hdl, "zpool_vdev_offline '%s'", devname);
+			zpool_vdev_offline(zhp, devname, B_TRUE);
+		} else if (!fmd_prop_get_int32(hdl, "spare_on_remove") ||
+		    replace_with_spare(hdl, zhp, vdev) == B_FALSE) {
```

Comment:
One alternative way to handle this would be to include a vdev's last expansion time in the config/vdev_tree.  Then the ZED could be updated to ignore "remove" event which immediately follow an expansion.  That's still not iron clad but it would cover the common case.

Reply:
Pushed some prototype code to ignore "remove" events for expanding vdevs.

My favourite Debian home builder cannot run "zpool_expand_001_pos" so this was tested very briefly on a "spare" CentOS7 builder:

```
/usr/share/zfs/test-runner/bin/test-runner.py  -c /var/tmp/zfs-tests.498.run -T functional -i /usr/share/zfs/zfs-tests -I 1
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/setup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/zpool_expand_001_pos (run as root) [00:13] [FAIL]
[   94.158494] BUG: unable to handle kernel paging request at ffffffffa0929000
[   94.159577] IP: [<ffffffffa0000fb4>] scsi_device_put+0x14/0x60 [scsi_mod]
[   94.160009] PGD 1816067 PUD 1817063 PMD d432b067 PTE 0
[   94.160009] Oops: 0000 [#1] SMP 
[   94.160009] Modules linked in: sd_mod crc_t10dif crct10dif_generic loop zfs(PO) icp(PO) zlua(PO) zcommon(PO) zunicode(PO) znvpair(PO) zavl(PO) spl(O) nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc crc32_pclmul aesni_intel aes_x86_64 lrw gf128mul glue_helper ablk_helper cryptd ppdev psmouse pcspkr serio_raw virtio_console i2c_piix4 evdev i2c_core processor parport_pc joydev parport pvpanic thermal_sys button virtio_balloon autofs4 ext4 crc16 mbcache jbd2 hid_generic usbhid hid sg sr_mod cdrom ata_generic virtio_scsi virtio_blk virtio_net crct10dif_pclmul crct10dif_common crc32c_intel ahci libahci floppy ata_piix xhci_hcd usbcore usb_common virtio_pci virtio_ring virtio libata scsi_mod [last unloaded: scsi_debug]
[   94.160009] CPU: 0 PID: 3544 Comm: systemd-udevd Tainted: P           O  3.16.0-4-amd64 #1 Debian 3.16.51-3
[   94.160009] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[   94.160009] task: ffff8800986ed450 ti: ffff8800d92dc000 task.ti: ffff8800d92dc000
[   94.160009] RIP: 0010:[<ffffffffa0000fb4>]  [<ffffffffa0000fb4>] scsi_device_put+0x14/0x60 [scsi_mod]
[   94.160009] RSP: 0018:ffff8800d92dfe68  EFLAGS: 00010292
[   94.160009] RAX: ffffffffa0929000 RBX: ffff880098480800 RCX: 0000000000000129
[   94.160009] RDX: 0000000000000000 RSI: 000000000002005d RDI: ffff880098473800
[   94.160009] RBP: ffff880098473800 R08: ffffffff81624a60 R09: 0000000000000000
[   94.160009] R10: ffff8800d9f3ed10 R11: 0000000000000246 R12: ffff880098480400
[   94.160009] R13: ffff8800d44d0ad8 R14: 000000000002005d R15: 0000000000000000
[   94.160009] FS:  00007f26f05fd880(0000) GS:ffff88011fc00000(0000) knlGS:0000000000000000
[   94.160009] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   94.160009] CR2: ffffffffa0929000 CR3: 0000000098023000 CR4: 00000000000006f0
[   94.160009] Stack:
[   94.160009]  ffff880098480800 ffff880098473800 ffffffffa09337cb ffff8800d44d0ac0
[   94.160009]  0000000000000010 ffffffff811e4f7d ffff8800d9f3ed00 0000000000000010
[   94.160009]  ffff8800d8f9cd58 ffff88011a662960 ffff8800d44bb018 ffff8800d8f9cd58
[   94.160009] Call Trace:
[   94.160009]  [<ffffffffa09337cb>] ? scsi_disk_put+0x2b/0x40 [sd_mod]
[   94.160009]  [<ffffffff811e4f7d>] ? __blkdev_put+0x15d/0x1a0
[   94.160009]  [<ffffffff811e59d1>] ? blkdev_close+0x21/0x30
[   94.160009]  [<ffffffff811b0dda>] ? __fput+0xca/0x1d0
[   94.160009]  [<ffffffff8108838c>] ? task_work_run+0x8c/0xb0
[   94.160009]  [<ffffffff81014ef9>] ? do_notify_resume+0x69/0xa0
[   94.160009]  [<ffffffff81524fca>] ? int_signal+0x12/0x17
[   94.160009] Code: 5b 5d 41 5c 41 5d 41 5e 41 5f c3 66 66 66 2e 0f 1f 84 00 00 00 00 00 e8 fb 61 52 e1 55 48 89 fd 53 48 8b 07 48 8b 80 c0 00 00 00 <48> 8b 18 48 85 db 74 0d 48 89 df e8 dc a0 0d e1 48 85 c0 75 17 
[   94.160009] RIP  [<ffffffffa0000fb4>] scsi_device_put+0x14/0x60 [scsi_mod]
[   94.160009]  RSP <ffff8800d92dfe68>
[   94.160009] CR2: ffffffffa0929000
[   94.160009] ---[ end trace ed75d0bd63d1621a ]---
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/cleanup (run as root) [00:00] [PASS]
```